packages/store/src/patterns/patternMatchers.js

```diff
@@ -539,27 +543,51 @@ const makePatternKit = () => {
    */
 
   /**
-   * @param {Passable} specimen
-   * @returns {Kind}
-   */
-  const kindOf = specimen => {
-    const style = passStyleOf(specimen);
-    if (style === 'tagged') {
-      return getTag(specimen);
-    } else {
-      return style;
-    }
-  };
-
-  /**
+   * Checks only recognized kinds, and only if the specimen
+   * passes the invariants associated with that recognition.
+   * If the kind is unrecognized, then we must conservatively assume
+   * that it does not satisfy invariants we cannot validate.
+   *
    * @param {Passable} specimen
    * @param {Kind} kind
-   * @param {Checker} [check]
+   * @param {Checker} check
    */
-  const checkKind = (specimen, kind, check = x => x) => {
-    const specimenKind = kindOf(specimen);
-    if (specimenKind === kind) {
-      return true;
+  const checkKind = (specimen, kind, check) => {
+    assert(
+      kind !== 'tagged',
+      X`"tagged" is used as a kind escape and cannot be used as a kind`,
+    );
+    /** @type {Kind} */
+    let specimenKind = passStyleOf(specimen);
+    if (specimenKind !== 'tagged') {
+      if (specimenKind === kind) {
+        // passStyleOf already did all the invariant checking.
+        return true;
+      }
+    } else {
+      specimenKind = getTag(specimen);
```

Comment:
this is new logic I don't see noted in the PR description

Reply:
Good catch. Done.